### PR TITLE
*: make txnCommitBatchSize adjustable by SQL

### LINKED ISSUE #33508: Make `txnCommitBatchSize` adjustable by config
## Enhancement

`txnCommitBatchSize` is a fixed value that cannot be adjusted, and make it could be easily adjusted and get the best configuration.

### PR DESCRIPTION
**Pull request by jackysp**
Signed-off-by: Jack Yu <jackysp@gmail.com>

### What problem does this PR solve?

Issue Number: close #33508

Problem Summary:
Expose `txnCommitBatchSize` to make it adjustable and use it to facilitate some performance optimization to find the best performance configuration.

### What is changed and how it works?
Set `tidb_txn_commit_batch_size` to 1M and run sysbench to prepare 64 tables of 10 million, the data import time is significantly reduced (the add indexing part is not included here.). The benefit here should come from a higher 1pc ratio.

master: 40min
截屏2022-03-28 17 42 59

this PR: 27min
截屏2022-03-26 11 14 44

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Add a variable `tidb_txn_commit_batch_size` to control the transaction commit batch size to make it easier to find the most reasonable value.
```

### CODE REVIEW

**Comment by jackysp** [comment]
The build should pass after the tidb pr merge corresponding to https://github.com/tikv/client-go/pull/456.

**Comment by morgo** [comment]
Is it possible to make this a sysvar (global) instead? When I speaked to @easonn7, his preference was to default to global for simplified configuration. We can use config when it refers more specifically to the resources (or paths) of a local machine.

**Comment by jackysp** [comment]
> Is it possible to make this a sysvar (global) instead? When I speaked to @easonn7, his preference was to default to global for simplified configuration. We can use config when it refers more specifically to the resources (or paths) of a local machine.

Global variables differ from configuration in two ways.

1. the configuration can be set separately for a particular tidb-server, which you also mentioned.
2. the configuration is truly global/cluster, either by restarting or through the `set config` syntax.

I care more about the second point, if you want to make global changes online, then the real point is to let all the established sessions benefit, but global variables do not support this semantically.

@morgo

**Comment by morgo** [comment]
> 2\. the configuration is truly global/cluster, either by restarting or through the `set config` syntax.
> 
> I care more about the second point, if you want to make global changes online, then the real point is to let all the established sessions benefit, but global variables do not support this semantically.

Global variables support online changes. The API supports a 'SetGlobal' func where you can specify the code to run:

https://github.com/pingcap/tidb/blob/09edaee061eff9099a103c23e6893ef4c0ed9e9a/sessionctx/variable/sysvar.go#L422-L435

Currently this code only runs on the node that sets the variable (although we might change this in future). To run the code on all nodes, you can modify here:

https://github.com/pingcap/tidb/blob/09edaee061eff9099a103c23e6893ef4c0ed9e9a/domain/sysvar_cache.go#L167-L247

**Comment by jackysp** [comment]
Interesting, I see it still uses the set global syntax. So, some of our server configurations were set with set session, but now the cluster level is set with set global as well? I could change it according to the same rules. It's just a little strange.

By the way, is the set config feature deprecated? @morgo

**Comment by jackysp** [comment]
related to https://github.com/tikv/client-go/pull/458

**Comment by jackysp** [comment]
image
I also tested it manually and manually changed the `tidb_txn_commit_batch_size` to larger at 15:41 minutes, and the 1pc usage rate improved.

**Comment by morgo** [comment]
> Interesting, I see it still uses `set global` syntax. So, some of tidb-server configurations were set with `set session`, but now the cluster level is set with `set global` as well? I could change it according to the same rules. It's just a little strange.

See https://github.com/pingcap/tidb/blob/master/docs/design/2021-12-08-instance-scope.md -- instance scoped variables traditionally used 'set session', but the plan is to use 'set global' instead. 'set session' is weird to MySQL users.

The initial framework for instance scope merged in https://github.com/pingcap/tidb/pull/32888 -- we are now working through the variables and converting them.

> By the way, is `set config` feature deprecated? @morgo

The goal of instance scope is that each tidb setting is viewable/settable with sysvars. But it's not a complete replacement for `set config`, because `set config` can change `pd` and `tikv`.

File: sessionctx/variable/sysvar.go [hunk 1]
```diff
@@ -529,6 +529,14 @@ var defaultSysVars = []*SysVar{
...
 		return nil
 	}},
+	{Scope: ScopeGlobal, Name: TiDBTxnCommitBatchSize, Value: strconv.FormatUint(tikvstore.DefTxnCommitBatchSize, 10), Type: TypeUnsigned, MinValue: 1, MaxValue: math.MaxInt64,
+		GetGlobal: func(sv *SessionVars) (string, error) {
+			return strconv.FormatUint(tikvstore.TxnCommitBatchSize.Load(), 10), nil
+		},
+		SetGlobal: func(s *SessionVars, val string) error {
+			tikvstore.TxnCommitBatchSize.Store(uint64(TidbOptInt64(val, int64(tikvstore.DefTxnCommitBatchSize))))
+			return nil
+		}},
```
[the remark this answers is not in the document]
Comment (jackysp) [reply sessionctx/variable/sysvar.go]: Addressed, PTAL @morgo , thx.

**Review by morgo — APPROVED** [review]

**Review by youjiali1995 — APPROVED** [review]

**Pull request merged**